### PR TITLE
Add resource document to google_firestore

### PR DESCRIPTION
**Pull request by ScottSuarez**
Added resource `document` to `google_firestore`
Closes [3950](https://github.com/hashicorp/terraform-provider-google/issues/3950)

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_firestore_document`
```

### CODE REVIEW

File: products/firestore/api.yaml [hunk 1]
```diff
@@ -108,3 +108,80 @@ objects:
...
+          Field names matching the regular expression __.*__ are reserved.
+          Reserved field names are forbidden except in certain documented contexts.
+          The map keys, represented as UTF-8, must not exceed 1,500 bytes and cannot be empty.
+
+          Field paths may be used in other contexts to refer to structured fields defined here.
+          For mapValue, the field path is represented by the simple or quoted field names of the
+          containing fields, delimited by .. For example, the structured field "foo" : { mapValue: { "x&y" : { stringValue: "hello" }}}
+          would be represented by the field path foo.x&y.
+
+          Within a field path, a quoted field name starts and ends with ` and may contain any character.
```
Comment (ScottSuarez) [inline products/firestore/api.yaml:176]: debating just linking to the documentation here... thoughts @rileykarson

Comment (rileykarson) [reply products/firestore/api.yaml:176]: Yeah, I'd consider just linking the docs here. What's interesting is that this is a JSON string, and that gets obscured by all these rules.

File: templates/terraform/decoders/firestore_document.go.erb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<%# The license inside this block applies to this file.
+	# Copyright 2019 Google Inc.
```
Comment (rileykarson) [inline templates/terraform/decoders/firestore_document.go.erb:2]:
```suggestion
	# Copyright 2020 Google Inc.
```

File: products/firestore/api.yaml [hunk 3]
```diff
@@ -108,3 +108,80 @@ objects:
                 be specified.
               values:
                 - :CONTAINS
+  - !ruby/object:Api::Resource
+    name: 'Document'
+    base_url: projects/{{project}}/databases/{{database}}/documents/{{collection}}
+    create_url: projects/{{project}}/databases/{{database}}/documents/{{collection}}?documentId={{document_id}}
+    update_verb: :PATCH
```
Comment (rileykarson) [inline products/firestore/api.yaml:115]: Can you add a test that performs an update?

File: templates/terraform/examples/firestore_document_nested_document.tf.erb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+resource "google_firestore_document" "<%= ctx[:primary_resource_id] %>" {
+  collection = "somenewcollection"
+  document_id = "my-doc-%{random_suffix}"
+  fields     = "{\"something\":{\"mapValue\":{\"fields\":{\"akey\":{\"stringValue\":\"avalue\"}}}}}"
+}
+
+resource "google_firestore_document" "sub_document" {
```
Comment (rileykarson) [inline templates/terraform/examples/firestore_document_nested_document.tf.erb:7]: Can you add an extra level of nesting? Either here or in another test.

Comment (ScottSuarez) [reply templates/terraform/examples/firestore_document_nested_document.tf.erb:7]: Sure !

File: products/firestore/api.yaml (hunk 3 above)
Comment (ScottSuarez) [reply products/firestore/api.yaml:115]: will do !

**Review by rileykarson — APPROVED** [review]

File: templates/terraform/decoders/firestore_document.go.erb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	# See the License for the specific language governing permissions and
+	# limitations under the License.
+-%>
+// metadata is not present if the API returns an error
+if name, ok := res["name"]; ok {
+	re := regexp.MustCompile("^projects/[^/]+/databases/[^/]+/documents/(.+)$")
+	match := re.FindStringSubmatch(name.(string))
+  if len(match) > 0{
+		res["path"] = match[1]
```
Comment (rileykarson) [inline templates/terraform/decoders/firestore_document.go.erb:20]: Can you comment somewhere inside the decoder that we use the decoder to add the path field?

**Pull request merged**